Add support for installing TLJH on Arm64 systems

cdibble opened the pull request:
Fixes [https://github.com/jupyterhub/the-littlest-jupyterhub/issues/62#](https://github.com/jupyterhub/the-littlest-jupyterhub/issues/62#)
Fixes #727 

 - [X ] Add / update documentation
 -  This PR shouldn't change anything from the user's point of view, but I added a couple of lines to state explicitly that this supports ARM 64 and AMD64 architectures and that it uses a miniforge conda environment.
 - [ X] Add tests
 - Updated ``tests/test_conda.py`` to use the miniforge installer instead of miniconda.

 <!-- Read more about our code-review guidelines at https://the-littlest-jupyterhub.readthedocs.io/en/latest/contributing/code-review.html -->

meeseeksmachine commented:
This pull request has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/tljh-on-m1-mac-arm-docker-installer-is-x86-specific/10894/2

consideRatio commented on tljh/utils.py:
```diff
@@ -51,5 +51,5 @@ def get_plugin_manager():
     pm.add_hookspecs(hooks)
     pm.load_setuptools_entrypoints('tljh')
-
+    
```
```suggestion

```

consideRatio commented on tljh/traefik.py:
```diff
@@ -40,7 +50,7 @@ def fatal_error(e):
 )
 def ensure_traefik_binary(prefix):
-    """Download and install the traefik binary"""
+    """Download and install the traefik binary-- prefix = '/opt/tljh/hub/'"""
```
```suggestion
    """Download and install the traefik binary to a location identified by a prefix path such as '/opt/tljh/hub/'"""
```

consideRatio commented on tljh/traefik.py:
```diff
@@ -13,12 +13,22 @@
-plat = "linux-amd64"
-traefik_version = "1.7.18"
+# traefik 2.7.x is not supported. Use v1.7.x
+# plat = "linux-amd64"
+# traefik_version = "1.7.18"
```
```suggestion
```

consideRatio commented on tljh/installer.py:
```diff
@@ -190,8 +197,9 @@ def ensure_user_environment(user_requirements_txt_file):
-        installer_url = "https://github.com/conda-forge/miniforge/releases/download/{v}/Mambaforge-{v}-Linux-x86_64.sh".format(v=mambaforge_new_version)
+        installer_url = "https://github.com/conda-forge/miniforge/releases/download/{v}/Mambaforge-{v}-Linux-{arch}.sh".format(v=mambaforge_new_version, arch=os.uname().machine)
         with conda.download_miniconda_installer(installer_url, installer_sha256) as installer_path:
+            logging.info(installer_path)
```
```suggestion
```

consideRatio commented on tljh/conda.py:
```diff
@@ -57,7 +57,7 @@ def download_miniconda_installer(installer_url, sha256sum):
 
         if sha256_file(f.name) != sha256sum:
-            raise Exception('sha256sum hash mismatch! Downloaded file corrupted')
+           raise Exception('sha256sum hash mismatch! Downloaded file corrupted')
```
```suggestion
            raise Exception('sha256sum hash mismatch! Downloaded file corrupted')
```

consideRatio commented on tests/test_conda.py:
```diff
@@ -24,7 +43,6 @@ def prefix():
         ])
         yield tmpdir
 
```
```suggestion


```

consideRatio commented on tests/test_conda.py:
```diff
@@ -8,14 +8,33 @@
 import tempfile
 
 
+# @pytest.fixture(scope='module')
+# def prefix():
+#     """
+#     Provide a temporary directory with a conda environment
+#     """
+#     miniconda_version = '4.7.10'
+#     miniconda_installer_sha256 = "8a324adcc9eaf1c09e22a992bb6234d91a94146840ee6b11c114ecadafc68121"
+#     installer_url = "https://repo.continuum.io/miniconda/Miniconda3-{}-Linux-x86_64.sh".format(miniconda_version)
+#     with tempfile.TemporaryDirectory() as tmpdir:
+#         with conda.download_miniconda_installer(installer_url, miniconda_installer_sha256) as installer_path:
+#             conda.install_miniconda(installer_path, tmpdir)
+#         conda.ensure_conda_packages(tmpdir, [
+#             'conda==4.8.1'
+#         ])
+#         yield tmpdir
+
```
```suggestion
```

consideRatio commented on setup.py:
```diff
@@ -4,7 +4,7 @@
     version='0.1',
     description='A small JupyterHub distribution',
-    url='https://github.com/jupyterhub/the-littlest-jupyterhub',
+    url='https://github.com/cdibble/the-littlest-jupyterhub',
```
```suggestion
    url='https://github.com/jupyterhub/the-littlest-jupyterhub',
```

consideRatio commented on docs/topic/installer-actions.rst:
```diff
@@ -30,7 +30,7 @@ of `traefik <http://traefik.io/>`_. This virtual environment is completely manag
 ================
 
-By default, a ``miniconda`` environment is installed in ``/opt/tljh/user``. This contains
+By default, a ``miniforge`` conda environment is installed in ``/opt/tljh/user``. This contains
```
```suggestion
By default, a ``mambaforge`` conda environment is installed in ``/opt/tljh/user``. This contains
```

consideRatio commented on docs/index.rst:
```diff
@@ -17,7 +17,7 @@ might still make breaking changes that have no clear upgrade pathway.
 ============
 
-The Littlest JupyterHub (TLJH) can run on any server that is running at least
+The Littlest JupyterHub (TLJH) can run on any server (x86_64/AMD64 and aarch64/ARM64 architectures are supported) that is running at least
 **Ubuntu 18.04**. Earlier versions of Ubuntu are not supported.
```
```suggestion
The Littlest JupyterHub (TLJH) can run on any server that is running **Ubuntu 18.04** or **Ubuntu 20.04** on a amd64 or arm64 CPU architecture. Earlier versions of Ubuntu are not supported.
```

consideRatio commented on bootstrap/bootstrap.py:
```diff
@@ -357,6 +357,13 @@ def serve_forever(server):
         logger.info('Upgrading TLJH installer...')
     run_subprocess(tljh_install_cmd)
 
+    pip_flags = ['--upgrade']
+    if os.environ.get('TLJH_BOOTSTRAP_DEV', 'no') == 'yes':
+        pip_flags.append('--editable')
+    tljh_repo_path = os.environ.get(
+        'TLJH_BOOTSTRAP_PIP_SPEC',
+        'git+https://github.com/jupyterhub/the-littlest-jupyterhub.git'
+    )
```
```suggestion
```

consideRatio commented on .circleci/config.yml:
```diff
@@ -0,0 +1,149 @@
+version: 2.1
```
```suggestion
# We use circleci to test our support for the Arm64 architecture. This file should mimic
# what is done in the GitHub workflows.
#
version: 2.1
```

consideRatio reviewed (changes requested):
@cdibble are you okay with me cleaning up the git history of this PR a bit and applying the changes suggested in my comments?

cdibble commented:
Hey @consideRatio - yes thank you for picking up the slack. I did bumble the git history badly didn't I. My apologies- a result of having rolled back a bunch of the changes in this original PR and then add some new changes from there. My own attempt to squash was ugly and I was too distracted with work projects to re-do it last week. So my apologies for that and thank you for making those changes.

Would you like me to go ahead and commit those suggestions to the PR?

consideRatio commented:
@cdibble I've squashed this PR into a few commits now, does this look okay to you?

I refactored some indentation etc as well in circleci config. I have a backup available of the previous PR state, and you probably have it as well on your local computer.

cdibble commented:
> @cdibble I've squashed this PR into a few commits now, does this look okay to you?
> 
> I refactored some indentation etc as well in circleci config. I have a backup available of the previous PR state, and you probably have it as well on your local computer.

LGTM. Thanks for doing that @consideRatio. You could drop that commit [08d937b](https://github.com/jupyterhub/the-littlest-jupyterhub/pull/679/commits/08d937b9902d7d947e9426fe04cf1366edbeefaa) and go back to the TLJH default of `tries=20` -- I had only bumped that up to help out with some dev and I didn't really mean for it to sneak into the PR. Of course, it doesn't really hurt anything, but in most cases where things actually work, `tries=20` is plenty.

consideRatio commented:
> but in most cases where things actually work, tries=20 is plenty.

Ah I wondered about that! Okay thanks for clarifying this @cdibble, and thanks for your work with this PR, thanks for sticking with it even though we were very low on maintenance capacity for a while in this repo! :heart: :tada:

cdibble commented:
Awesome @consideRatio - I am glad to be of some help and I much appreciate this project and all of its contributors!

My apologies for the added overhead here- my own need to refresh my knowledge of what was going on led to some fumbling.

consideRatio commented:
Wieee! Thanks again @cdibble!!!! :heart: :tada:
